app/javascript/react/components/ReviewForm.js

```diff
@@ -0,0 +1,139 @@
+import React, { useState } from "react"
+import ErrorDisplay from "./ErrorDisplay"
+
+const ReviewForm = (props) => {
+  const [newReview, setNewReview] = useState ({
+    title:"",
+    rating:"",
+    heatIndex:"",
+    body:""
+  })
+  const [errors, setErrors] = useState({})
+
+  const postReview = async (formData) =>{
+    
+    try{
+      const response = await fetch(`/api/v1/sauces/${props.sauce.id}/reviews`, {
+        credentials: "same-origin",
+        method: "POST",
+        headers: {
+          "Accept": "application/json",
+          "Content-Type": "application/json"
+        },
+        body: JSON.stringify({review: formData})
+      })
+      if (!response.ok) {
+        const errorMessage = `${response.status} (${response.status.text})`
+        const error = new Error(errorMessage)
+        throw(error)
+      }
+      const reviewObject = await response.json()
+      appendNewReview(reviewObject.review)
+
+    } catch(error) {
+      console.log(`Error in fetch: ${error}`)
+    }
+  }
+
+  const handleChange = (event) => {
+    setNewReview ({
+      ...newReview,
+      [event.currentTarget.name]: event.currentTarget.value
+    })
+  }
+
+  const validForSubmission = () => {
+    let submitErrors = {}
+    const requiredFields = ["title", "rating", "heatIndex", "body"]
+    requiredFields.forEach(field => {
+      if (newReview[field].trim() == "") {
+        submitErrors = {
+          ...submitErrors,
+          [field]: "cannot be blank"
+        }
+      }
+      setErrors(submitErrors)
+    })
+    return errors
+  }
+
+  const handleFormSubmit = (event) => {
+    event.preventDefault()
+    
+    if (validForSubmission()) {
+      postReview(newReview)
+      setNewReview ({
+        title:"",
+        rating:"",
+        heatIndex:"",
+        body:""
+      })
+    }
+  }
+
+  const appendNewReview = (reviewPayload) => {
+    const newReviews = props.sauce.reviews
+    newReviews.push(reviewPayload)
+    const sauceWithNewReview = {
+      ...props.sauce,
+      reviews: newReviews
+    }
+    props.setSauce(sauceWithNewReview)
+  }
```

Comment:
this doesnt need to be its own method if we dry it up! 

```
props.setSauce({
    ...props.sauce,
    reviews: [...props.sauce.reviews, reviewPayload
})
```
which can be called in your fetch function

Reply:
Fixed